File: go/data_type.go
/*
자료형이란 데이터를 표현하는 기준입니다.
go의 자료형은 논리형 / 문자형 / 정수형 / 실수형 / 복소수형 / 룬형 / 바이트형이 있습니다.
*/

package main

import (
	"fmt"
	"unsafe"
)

func main() {
	/*
	   - 정수형 -
	   int8: 1byte (-128 ~ 127)
	   int16: 2byte (-32768 ~ 32767)
	   int32: 4byte (-2147483648 ~ 2147483647)
	   int64: 8byte (-9223372036854775808 ~ 9223372036854775807)
	   uint8: 1byte (0 ~ 255)
	   uint16: 2byte (0 ~ 65535)
	   uint32: 4byte (0 ~ 4294967295)
	   uint64: 8byte (0 ~ 18446744073709551615)
	   uintptr: pointer를 저장하는 자료형입니다. (uint와 크기가 같습니다.)

	   uint:32bit에서는 uint32, 64bit에서는 uint64입니다.
	   int:32bit에서는 int32, 64bit에서는 int64입니다.
	*/
	fmt.Println("\n=== 정수형 ===")
	var in int = 3
	fmt.Println("in은", in, "byte는", unsafe.Sizeof(in))
	var _in int64 = 3
	fmt.Println("_in은", _in, "byte는", unsafe.Sizeof(_in))

	/*
	   - 실수형 -
	   float32: 4byte (? ~ 3.4E38) - ieee-754 32bit 단정밀도 부동소수점
	   float64: 8byte (?~ 1.7E308) - ieee-754 32bit qo정밀도 부동소수점
	*/
	fmt.Println("\n=== 실수형 ===")
	var fn float32 = 3.0
	fmt.Println("fn은", fn, "byte는", unsafe.Sizeof(fn))
	var dn float64 = 3.0
	fmt.Println("dn은", dn, "byte는", unsafe.Sizeof(dn))

	/*
	   - 복소수형 -
	   complex64: 4byte (?) float32의 실수부와 허수부로 이루어진 자료형입니다.
	   complex128: 8byte (?) float64의 실수부와 허수부로 이루어진 자료형입니다.
	*/
	fmt.Println("\n=== 복소수형 ===")
	var ci complex64 = 3 + 5i
	fmt.Println("ci은", ci, "실수부는", real(ci), "허수부는", imag(ci), "byte는", unsafe.Sizeof(ci))

	/*
	   - 문자형 -
	   rune: unicode(utf-8)를 저장하는 자료형입니다. (int32와 크기가 같습니다.)
	   문자는 ''로 만듭니다.
	*/
	fmt.Println("\n=== 복소수형 ===")
	var ch rune = 'g'
	fmt.Println("ch는", ch, "byte는", unsafe.Sizeof(ch))
	var chn rune = 103
	fmt.Println("chn은", chn, "byte는", unsafe.Sizeof(chn))

	fmt.Println("ch == chn은", ch == chn)

	/*
	   - byte형 -
	   byte에 저장되어 있는 자료형입니다.
	*/
	fmt.Println("\n=== byte형 ===")
	var bi byte = 103
	fmt.Println("bi는", bi, "byte는", unsafe.Sizeof(bi))
	var bn byte = 'g'
	fmt.Println("bn은", bn, "byte는", unsafe.Sizeof(bn))

	/*
	   - 형변환 -
	   연산을 위해 자료형을 일치시켜야 하는 때가 있습니다. 이를 위해 변수 및 리터럴의 자료형을 다른 자료형으로 바꾸는것을 형변환이라 합니다.
	   go는 strong type 언어로 자동형변환이 일어나지 않습니다. (<- 대입연산은 아니지만, 산술연산에서는 일어나는듯)
	*/
	fmt.Println("\n=== 형변환 ===")
	var ia int32 = 3
	var ib int64 = 5
	// fmt.Println("ia + ib는", ia + ib) - error: ia와 ib의 자료형이 일치하지 않습니다.
	fmt.Println("ia + ib는", ia+int32(ib), "byte는", unsafe.Sizeof(ia+int32(ib)))
	fmt.Println("ia + ib는", int64(ia)+ib, "byte는", unsafe.Sizeof(int64(ia)+ib))

	var ra float32 = 3.0
	var rb float64 = 5.0
	// fmt.Println("ra + rb는", ra + rb) - error: ra와 rb의 자료형이 일치하지 않습니다.
	fmt.Println("ra + rb는", ra+float32(rb), "byte는", unsafe.Sizeof(ra+float32(rb)))
	fmt.Println("ra + rb는", float64(ra)+rb, "byte는", unsafe.Sizeof(float64(ra)+rb))

	var ic int64 = 3 / 2
	var rc float64 = 3 / 2
	var rd float64 = float64(3) / 2
	fmt.Println("ic는", ic, "byte는", unsafe.Sizeof(ic))
	fmt.Println("rc는", rc, "byte는", unsafe.Sizeof(rc))
	fmt.Println("rd는", rd, "byte는", unsafe.Sizeof(rd))
	fmt.Println("int64(rd)는", int64(rd), "byte는", unsafe.Sizeof(rd))

	/*
	   - 상수 -
	   const로 상수를 선언합니다.
	*/
	fmt.Println("\n=== 상수 ===")
	const M int = 5
	fmt.Println("M은", M, "byte는", unsafe.Sizeof(M))

	/*
			   - 짧은선언 -
		       :=를 통해 var를 사용하지 않고 변수를 선언할 수 있습니다.
	           이때 자료형은 오른쪽의 자료형으로 정의됩니다.
	*/
	fmt.Println("\n=== 짧은선언 ===")
	m := 3
	fmt.Println("m은", m, "byte는", unsafe.Sizeof(m))

	/*
			    - 변수 비활성화 -
		        go는 사용하지 않는 변수 및 패키지가 있으면 compile error를 띄웁니다.
		        이때 _ 를 통해 변수를 비활성화할 수 있습니다.
	*/
	fmt.Println("\n=== 변수 비활성화 ===")
	var a int = 3
	_ = a

	/*
	   - scope -
	   scope는 변수의 유효범위로, go는 block scope입니다.
	*/
	fmt.Println("\n=== scope ===")
	var p int = 3
	{
		var p int = 5
		fmt.Println("p는", p, "byte는", unsafe.Sizeof(p))

	}
	fmt.Println("p는", p, "byte는", unsafe.Sizeof(p))

	/*
		   - literal 자료형 -
	literal의 자료형은 int, float64, bool입니다.
	*/
	fmt.Println("\n=== scope ===")
	fmt.Printf("3의 자료형은 %T입니다.\n", 3)
	fmt.Printf("3.0의 자료형은 %T입니다.\n", 3.0)
	fmt.Printf("3 / 2의 자료형은 %T입니다.\n", 3/2)
	fmt.Printf("3.0 / 2의 자료형은 %T입니다.\n", 3.0/2)
	fmt.Printf("true의 자료형은 %T입니다.\n", true)
	fmt.Printf("'g'의 자료형은 %T입니다.\n", 'g')
	fmt.Printf("\"golang\"의 자료형은 %T입니다.\n", "golang")
	fmt.Printf("byte(3)의 자료형은 %T입니다.\n", byte(3))
	fmt.Printf("3 + 5i의 자료형은 %T입니다.\n", 3+5i)

	// 📝 3.0 / 2는 float64와 int의 연산이지만 float64로 연산되는 것을 보면, go는 변수에 대해 strong type 언어인듯 보입니다.
	fmt.Printf("3.0의 자료형은 %T, 2의 자료형은 %T, 3.0 / 2의 자료형은 %T입니다.\n", 3.0, 2, 3.0/2)
}

File: c/data_type.c
/*
자료형이란 데이터를 표현하는 기준입니다.
c의 자료형은 정수형 / 실수형이 있습니다.
- 문자형(char)은 정수형으로 간주됩니다.
- 논리형은(1byte) 정수형으로 간주됩니다. (1:참, 0:거짓)
*/

#include <stdio.h>

int main(void)
{
  /*
  - 정수형-
  char: 1byte (-128 ~ 127)
  short: 2byte (-32768 ~ 32767)
  int: 4byte (-2147483648 ~ 2147483647)
  long: 4byte (-2147483648 ~ 2147483647)
  long long: 8byte (-9223372036854775808 ~ 9223372036854775807)
  ✏️ 일반적으로 int형을 사용합니다.
  */
  printf("=== 정수형 ===\n");
  int in = 3;
  // sizeof로 자료형의 byte를 알 수 있습니다.
  printf("in의 byte는 %lu \n", sizeof(in));

  // 정수형 연산은 short끼리 연산해도 int로 형변환됩니다. (int형이 연산이 빠르기 때문에)
  short ia = 3;
  short ib = 5;
  printf("ia, ib의 byte는 %lu, %lu \n", sizeof(ia), sizeof(ib));
  printf("ia + ib의 byte는 %lu \n", sizeof(ia + ib));

  // 접미사(u, l, ul, ll, ull)을 붙여 리터럴 자료형을 정의합니다. (붙이지 않으면 int형입니다.)
  long long lli = 5ll;
  printf("lli는 %lld %lu byte \n", lli, sizeof(lli));



  /*
  - 실수형 -
  float: 4byte (3.4E-37 ~ 3.4E38)
  double: 8byte (1.7E-307 ~ 1.7E308)
  long double: double보다 큰 자료형
  ✏️ 일반적으로 double형을 사용합니다.
  */
  printf("=== 실수형 ===\n");
  int rn = 3;
  printf("rn의 byte는 %lu \n", sizeof(rn));

  // 정수형 연산은 float끼리 연산해도 double로 형변환되지는 않습니다.
  float ra = 3.0;
  float rb = 5.0;
  printf("ra, rb의 byte는 %lu, %lu \n", sizeof(ra), sizeof(rb));
  printf("ra + rb의 byte는 %lu \n", sizeof(ra + rb));

  // 접미사(f, l)을 붙여 리터럴 자료형을 정의합니다. (붙이지 않으면 double형입니다.)
  float fr = 5.0f;
  printf("fr은 %f %lu byte \n", fr, sizeof(fr));

  double dr = 5.0l;
  printf("dr은 %f %lu byte \n", dr, sizeof(dr));

  /*
  - 문자형 -
  문자형은 char에 담깁니다.
  문자(literal)는 '', 문자열(literal)은 ""로 만듭니다.
   */
  printf("=== 문자형 ===\n");
  char ch = 'c';
  printf("ch의 byte는 %lu \n", sizeof(ch));
  printf("ch의 문자 및 unicode는 %c, %d \n", ch, ch);

  char nch = 99;
  printf("nch의 문자 및 unicode는 %c, %d \n", nch, nch);

  int ich = 99;
  printf("ich의 문자 및 unicode는 %c, %d \n", ich, ich);


  /*
  - 형변환 -
  연산을 위해 자료형을 일치시켜야 하는 때가 있습니다. 이를 위해 변수 및 리터럴의 자료형을 다른 자료형으로 바꾸는것을 형변환이라 합니다.
  형변환에는 자동(묵시적)형변환, 강제(명시적)형변환 두가지가 있습니다.
  */
  printf("=== 형변환 ===\n");
  // 자동(묵시적)형변환은 대입연산(=)연산에서 형이 다르면 일어납니다. 대입연산은 왼쪽 피연산자의 자료형으로 형변환됩니다.
  double dm = 3;
  printf("dm은 %f byte %lu \n", dm, sizeof(dm));

  int im = 3.5; // warning
  printf("im은 %d byte %lu \n", im, sizeof(im));

  // 자동(묵시적)형변환은 산술연산에서 형이 다르면 일어납니다. 산술연산은 넓은 자료형으로 형변환됩니다.
  double dk = 3 / 2;
  printf("dk는 %f byte %lu \n", dk, sizeof(dk));

  double dj = (double)3 / 2;
  printf("dj는 %f byte %lu \n", dj, sizeof(dj));


  /*
  - 상수 -
  const로 상수를 선언합니다.
  */
  printf("=== 상수 ===\n");
  const int M = 5;
  printf("M은 %d \n", M);

  /*
  - scope -
  scope는 변수의 유효범위로, c는 block scope입니다.
  */
  printf("=== scope ===\n");
  int s = 3;
  printf("s는 %d입니다.\n", s);

  {
    int s = 5;
    printf("s는 %d입니다.\n", s);
  }

  /*
  - static / register -
  */
}
